### PR TITLE
refactor: Add TranslationFacility for decoupling from HOVerwaltung

### PR DESCRIPTION
**Pull request by sgcr**
1. changes proposed in this pull request:
- In the PR #2141 for feature #2140 (ArenaInfoPanel) were some comments. So I started developing a `HumanDuration` class that could also produce a human readable string from itself.
That needed translations, and when I wrote tests, I noticed that I needed an instance of `HOVerwaltung`, which is a singleton. I got a strange error message that the DB is working and the reason is the singleton.
Those are the facts.
The idea is to have a holder of the globally used instance of the translator: `TranslationFacility`
The main advantage is that it is possible to change the translator of the application during the tests (e.g., the `LANGUAGE_NO_TRANSLATION` or `LANGUAGE_DEFAULT`) without doing anything else by accident.

2. `src/main/resources/release_notes.md` ...
 - [ ] has been updated

3. [Optional] suggested person to review this PR @tychobrailleur @wsbrenk

### CODE REVIEW

File: src/test/java/core/model/TranslatorTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import java.util.MissingResourceException;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.junit.jupiter.params.provider.Arguments.of;
+
+class TranslatorTest {
+
+    private static final List<String> SUPPORTED_LANGAUGES = List.of(
```
Comment (tychobrailleur) [inline src/test/java/core/model/TranslatorTest.java:21]: Typo, `LANGUAGES` (a few more occurrences in this file)

File: src/test/java/core/model/TranslationFacilityTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Test
+    void setLanguage_unknown_throws_exception() {
+        // when-then
+        assertThatThrownBy(() -> TranslationFacility.setLanguage("foobar")).isInstanceOf(MissingResourceException.class);
+    }
+
+    @Test
+    void setTranslator() {
```
Comment (tychobrailleur) [inline src/test/java/core/model/TranslationFacilityTest.java:43]: Do we really need this test, given `getTranslator` pretty much does the same thing?

File: src/test/java/core/model/TranslationFacilityTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        // when
+        final var translation = TranslationFacility.tr("ls.button.save");
+
+        // then
+        assertThat(translation).isEqualTo("!ls.button.save!");
+    }
+
+    @Test
+    void trWithVariables() {
```
Comment (tychobrailleur) [inline src/test/java/core/model/TranslationFacilityTest.java:65]: Here again we should strive for consistency, either use snake case everywhere, or camel case everywhere.

**Review by tychobrailleur — CHANGES_REQUESTED** [review]
Thanks!  Some minor changes and comments, but definitely a change that improves testability.  Also, please keep in mind that one of the project goals is to migrate to Kotlin.

File: src/main/java/core/model/TranslationFacility.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package core.model;
+
+import static java.util.Objects.requireNonNullElseGet;
+
+public class TranslationFacility {
+
+    private static final String INITIAL_TRANSLATOR = Translator.LANGUAGE_NO_TRANSLATION;
+
+    private static Translator translator;
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/TranslationFacility.java:9]: > I don't see the `Translator` class in this review, has this been added in a separate PR?

It was added previously here in #2108

It was a lighter decoupling (preparation) for this PR here.

File: src/main/java/core/model/HOVerwaltung.java [hunk 5]
```diff
@@ -90,9 +86,13 @@ public void setResource(String pfad) {
 		setTranslator(pfad);
 	}
 
+	/**
+	 * @deprecated Provided for compatibility. Please use {@link TranslationFacility#setLanguage(String)} instead.
+	 */
+	@Deprecated(since = "9.0", forRemoval = true)
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/HOVerwaltung.java:92]: Also possible but it might be difficult to merge from release branches when some refactoring is already in the master branch.

File: src/main/java/core/gui/model/ArenaStatistikTableModel.java [hunk 6]
```diff
@@ -5,19 +5,18 @@
...
-import java.awt.Color;
-import java.text.DateFormat;
+import javax.swing.*;
+import javax.swing.table.AbstractTableModel;
+import java.awt.*;
 import java.time.Duration;
 
-import javax.swing.SwingConstants;
-import javax.swing.table.AbstractTableModel;
+import static core.model.TranslationFacility.tr;
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/gui/model/ArenaStatistikTableModel.java:19]: A static import is shorter when you only read the `tr` in the code.
There is more oversight in the code and on the other side a missing `tr` can be overseen.
The single `tr` can searched with the `Find usage` functionality in IDEA but not so good with a normal text search.

On the other side calls to `TranslationFacility.tr` are still easy to find by a normal text search.
It is shorter than the old style with `HOVerwaltung.instance()...` but longer than the sole `tr`.
It is normally not overseen.

I prefer a single `tr` on new code and for existing code there was the question if it's possible to convert by a "Find an Replace".

Due to a lot of files to modify I had to do it pragmatically and I nevertheless it took a bunch of time.

So my suggestion is to use both variants due to advantages and disadvantages:
* `tr` with a static import
* `TranslationFacility.tr` with a normal import

But if it bothers you I am open for some proposals regarding the code style.

File: src/main/java/core/model/player/MatchRoleID.java [hunk 7]
```diff
@@ -193,27 +194,27 @@ public static String getShortNameForPosition(byte posId) {
...
+			case MIDFIELDER_TOWING -> getLangStr("ls.player.position_short.innermidfieldertowardswing");
+			case WINGER -> getLangStr("ls.player.position_short.winger");
+			case WINGER_TOMID -> getLangStr("ls.player.position_short.wingertowardsmiddle");
+			case WINGER_OFF -> getLangStr("ls.player.position_short.wingeroffensive");
+			case WINGER_DEF -> getLangStr("ls.player.position_short.wingerdefensive");
+			case FORWARD -> getLangStr("ls.player.position_short.forward");
+			case FORWARD_TOWING -> getLangStr("ls.player.position_short.forwardtowardswing");
+			case FORWARD_DEF -> getLangStr("ls.player.position_short.forwarddefensive");
+			case SUBSTITUTED1, SUBSTITUTED2, SUBSTITUTED3 -> getLangStr("Ausgewechselt");
+			default -> getLangStr("Unbestimmt");
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/player/MatchRoleID.java]: Review Finding 1

File: src/main/java/module/lineup/CopyListener.java [hunk 8]
```diff
@@ -95,13 +95,13 @@ public static void copyToClipboard(final String txt) {
...
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.midfield")).append(": ").append(lineupRatingPanel.getMidfieldRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.rightdefence")).append(": ").append(lineupRatingPanel.getRightDefenseRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.centraldefence")).append(": ").append(lineupRatingPanel.getCentralDefenseRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.leftdefence")).append(": ").append(lineupRatingPanel.getLeftDefenseRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.rightattack")).append(": ").append(lineupRatingPanel.getRightAttackRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.centralattack")).append(": ").append(lineupRatingPanel.getCentralAttackRating()).append(LF);
+			sb.append(TranslationFacility.tr("ls.match.ratingsector.leftattack")).append(": ").append(lineupRatingPanel.getLeftAttackRating()).append(LF);
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/module/lineup/CopyListener.java:104]: > Nothing to do here, but one day we will need to fix this appending of localised strings, this is wrong for certain languages. Also `StringBuilder` is probably no longer needed.

I completely agree with you.

You could create a small development bug/refactor issue and we can do it a little bit later.
But we should not mix these things here in the PR. ;-)

File: src/test/java/core/model/TranslationFacilityTest.java (hunk 2 above)
Comment (sgcr) [reply src/test/java/core/model/TranslationFacilityTest.java:43]: > Do we really need this test, given `getTranslator` pretty much does the same thing?

Yes, because one test is explicit for `getTranslator` and explicit for `setTranslator`.
For `setTranslator` is important that you cannot set it to `null`.
It will still return a translator which is the initial translator with no translation.

The test for `setTranslator` can be different.
I changed it.

Review Finding 2

File: src/main/java/core/model/TranslationFacility.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    static {
+        translator = Translator.load(INITIAL_TRANSLATOR);
+    }
+
+    private TranslationFacility() {
+        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/TranslationFacility.java:15]: > The Exception is probably overkill, and does not fit with the existing pattern.

It makes clear that an instance of the class is not supported there is no intention to create an instance because it is only provided as a namespace.

Think of a class that is a singleton (regardless the questions if singletons are a good idea) when the constructor is private normally.

The thrown exception marks the class clearly as a utility class that is no singleton.

File: src/test/java/core/model/TranslatorTest.java (hunk 1 above)
Comment (sgcr) [reply src/test/java/core/model/TranslatorTest.java:21]: Review Finding 3

File: src/main/java/core/net/DownloadDialog.java [hunk 10]
```diff
@@ -251,6 +251,7 @@ public void windowOpened(WindowEvent e) {
 			}
 		});
 
+		setLocationByPlatform(true);
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/net/DownloadDialog.java:254]: > What does that do?

Added by accident.

Will be added by another PR.

Will be removed.

[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/net/DownloadDialog.java]: Review Finding 4

File: src/main/java/core/module/config/ModuleConfigDialog.java [hunk 11]
```diff
@@ -55,10 +49,10 @@ private JPanel createButtons() {
         JPanel buttonPanel = new ImagePanel();
         ((FlowLayout) buttonPanel.getLayout()).setAlignment(FlowLayout.CENTER);
 
-        okButton = new JButton(HOVerwaltung.instance().getLanguageString("ls.button.save"));
+        okButton = new JButton(TranslationFacility.tr("ls.button.save"));
         okButton.addActionListener(this);
 
-//        cancelButton = new JButton(HOVerwaltung.instance().getLanguageString("ls.button.cancel"));
```
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/module/config/ModuleConfigDialog.java:61]: Review finding 6

File: src/main/java/core/model/HOVerwaltung.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/HOVerwaltung.java:92]: Not sure I get this.  `master` will be for a `9.0`, which is where the refactoring is ongoing, when the release branches would theoretically only get bug fixes.  Why would there be issues merging?  AFAIU this specific `setTranslator` method was introduced on `master` and `9.0` was not released so it should be fine to remove?

File: src/main/java/core/model/TranslationFacility.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/TranslationFacility.java:15]: OK, then if it's the new pattern for singletons in HO, this needs to be applied to all.  We are open to new styles and new approaches, but consistency in my mind is important.  Also, just to note this will naturally go away with Kotlin `object` declaration.

File: src/main/java/core/gui/model/ArenaStatistikTableModel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/gui/model/ArenaStatistikTableModel.java:19]: > Due to a lot of files to modify I had to do it pragmatically and I nevertheless it took a bunch of time.

Tell me about it, I had to review them all. 😆

> But if it bothers you I am open for some proposals regarding the code style.

Hmm, I think the bothering aspect is to sometimes have `tr`, sometimes `TranslationFacility.tr`.  Would you be happy to just use `TranslationFacility.tr`?  It seems to be the most commonly used version in the PR.

File: src/main/java/core/rating/RatingPredictionModel.java [hunk 12]
```diff
@@ -14,13 +14,18 @@
...
 import org.jetbrains.annotations.NotNull;
+
 import java.util.*;
+
 import static core.constants.player.PlayerSkill.*;
 import static core.model.match.IMatchDetails.*;
-import static core.model.player.IMatchRoleID.*;
 import static core.model.player.IMatchRoleID.KEEPER;
 import static core.model.player.IMatchRoleID.WINGER;
-import static java.lang.Math.*;
```
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/rating/RatingPredictionModel.java:23]: > Additionally code style is also an sole topic on its own...

Yes, and I definitely have no desire to go into that topic. 😊  I think the difference lies in static import and class import, but this is not important -- again that was more curiosity than anything else.

File: src/main/java/module/lineup/CopyListener.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/module/lineup/CopyListener.java:104]: Exactly, as I said nothing to be done.  I'll create the relevant ticket, thanks!

File: src/main/java/core/model/TranslationFacility.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/TranslationFacility.java:9]: Ah I see, ok, thanks!

File: src/main/java/core/model/HOVerwaltung.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/HOVerwaltung.java:92]: > Not sure I get this. `master` will be for a `9.0`, which is where the refactoring is ongoing, when the release branches would theoretically only get bug fixes. Why would there be issues merging? AFAIU this specific `setTranslator` method was introduced on `master` and `9.0` was not released so it should be fine to remove?

It depends on the branching model we are using.
What I saw is:
There is a release branch `release/8` to make the application stable.
The release branch is a branch from `master`.

Bugs are fixed in that branch `release/8` to get a usable product.
Then the release it built from this release branch.
Right?

The main question is **how** the fixes/changes in that release branch are merged into the `master` branch.
The master branch needs these changes normally because the `master` branch is the origin for next release branch.
When the master branch is not updated with these fixes/changes the bugs occur again when a release is created from `master` for the next release branch.

The normal GIT action to perform that update from the `master` branch is to merge from the current release branch that contains the fixes/changes.

**In the normal development process there are these refactoring like the PR here.**

So when you think of a fix of a line in the release branch that used the old API before this PR here (when resident in master) was merged, you will get compile time errors.
To prevent this you can still use the old and deprecated API here.
After the last release of version `9`, so when the next release branch `release/9` will not get any update, the deprecated API can be removed in master.

So basically the idea behind the annotation for deprecation is to help merging from release branches to the master branch.
This does not help for merge conflicts but if there are any you have the advantage to use parts of the old API here.

BTW:
I see no merge lines in the GIT tree.

For example here:
#2097 was fixed with 5c50df018408b1c4c6e41512fe5322faa2849159 in `release/8` branch
The fix was related to file 'MatchLineupDetails.java'
With #2100 and commit 51049cd0c262c8dbbe960b4649638b5af9a351fe the same thing was done in the `master` branch be you cannot see that in the GIT tree.

How are you creating these merges?

File: src/main/java/core/gui/model/ArenaStatistikTableModel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/gui/model/ArenaStatistikTableModel.java]: Review Finding 7

The shorter version with `tr` is clearly more readable.

File: src/main/java/core/model/TranslationFacility.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/TranslationFacility.java:15]: > OK, then if it's the new pattern for singletons in HO, this needs to be applied to all. We are open to new styles and new approaches, but consistency in my mind is important. Also, just to note this will naturally go away with Kotlin `object` declaration.

It is not a pattern for singletons but utility classes (namespaces in C++ e.g.).
A utility class is a class that provides static functions that are grouped together under the name of the class.
The class itself has a private constructor and unlike to singletons the private constructor is never invoked.
The exception is only to make that clear.

Singletons normally have a private constructor that is invoked in the `instance` function.

File: src/main/java/core/model/HOVerwaltung.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/HOVerwaltung.java:92]: Usually @wsbrenk takes care of the release branching, so I'll let him pitch in here.  But IMHO, this is making things more complicated than they need to be.  If you're merging, and you're getting conflicts, you fix at merge time.  I cannot recall a situation where this happened in HO, but @wsbrenk is probably better placed to comment.

File: src/main/java/core/model/TranslationFacility.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/TranslationFacility.java:15]: Sorry, that's what I was meaning to say, “non-instantiable utility classes” rather than “singleton.”  The consistency argument still stands, regardless of the incorrect terminology.

[the remark this answers is not in the document]
Comment (sgcr) [reply src/main/java/core/model/TranslationFacility.java:15]: No, problem.

Please note that the project is not consistent in several places. ;-)

I will create another small refactor PR if you want.

[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/TranslationFacility.java:15]: Yes, for sure. :-)
Thanks, Sebastian, appreciate the effort!!

File: src/main/java/core/model/HOVerwaltung.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tychobrailleur) [reply src/main/java/core/model/HOVerwaltung.java:92]: Right, I think I'll let it go to move ahead with that PR that's be open for a while, but let's reconsider that pattern.

**Review by tychobrailleur — APPROVED** [review]
Thanks for the work!
